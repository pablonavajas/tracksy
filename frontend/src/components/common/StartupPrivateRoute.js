import React from "react";
import { connect } from "react-redux";
import { Route, Redirect } from "react-router-dom";
import PropTypes from "prop-types";
import Preloader from "../layout/Preloader";

export const StartupPrivateRoute = ({
  component: Component,
  auth,
  ...rest
}) => (
  <Route
    {...rest}
    render={(props) => {
      if (auth.isLoading) {
        return <Preloader />;
      } else if (!auth.isAuthenticated && !localStorage.token) {
        return <Redirect to="/login" />;
      } else if (auth.isStartup === false && auth.isLoading === false) {
        return <Redirect to="/" />;
      } else {
        return <Component {...props} />;
      }
    }}
  />
);

const mapStateToProps = (state) => ({
  auth: state.auth,
});

export default connect(mapStateToProps)(StartupPrivateRoute);
