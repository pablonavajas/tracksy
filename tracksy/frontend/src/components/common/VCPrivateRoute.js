import React from "react";
import { connect } from "react-redux";
import { Route, Redirect } from "react-router-dom";
import PropTypes from "prop-types";

const VCPrivateRoute = ({ component: Component, auth, ...rest }) => (
  <Route
    {...rest}
    render={props => {
      if (auth.isLoading) {
        return (
          <div className="progress">
            <div className="indeterminate" />
          </div>
        );
      } else if (!auth.isAuthenticated && !localStorage.token) {
        return <Redirect to="/login" />;
      } else if (auth.isStartup) {
        // console.log(auth.isStartup);
        return <Redirect to="/startupForm" />;
      } else {
        return <Component {...props} />;
      }
    }}
  />
);

const mapStateToProps = state => ({
  auth: state.auth
});

export default connect(mapStateToProps)(VCPrivateRoute);
