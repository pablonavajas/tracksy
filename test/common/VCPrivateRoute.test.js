import React from "react";
import { shallow, mount } from "enzyme";
import { VCPrivateRoute } from "../../tracksy/frontend/src/components/common/VCPrivateRoute";
import ConnectedVCPrivateRoute from "../../tracksy/frontend/src/components/common/VCPrivateRoute";
import toJson from "enzyme-to-json";
import "../setupTests";
import { HashRouter as Router } from "react-router-dom";

//store needed for testing mapStateToProps
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
const mockStore = configureMockStore([thunk]);

//only VC Funds can register
describe("VCPrivateRoute", () => {
  let auth, wrapper;

  beforeEach(() => {
    auth = {
      isAuthenticated: true,
      isLoading: false,
      isStartup: false,
    };
  });

  it("shows preloader, if isLoading is true ", () => {
    auth.isLoading = true;
    wrapper = mount(
      <Router>
        <VCPrivateRoute auth={auth} />
      </Router>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("redirects to login, if isAuthenticated is false ", () => {
    auth.isAuthenticated = false;
    wrapper = mount(
      <Router>
        <VCPrivateRoute auth={auth} />
      </Router>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("redirects to /startupPage, if isAuthenticated is true, isStartup is true and isLoading is false", () => {
    auth.isStartup = true;
    wrapper = mount(
      <Router>
        <VCPrivateRoute auth={auth} />
      </Router>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

// Testing MapStateToProps
describe("VCPrivateRoute ", () => {
  let store, wrapper;

  beforeAll(() => {
    store = mockStore({ auth: { isAuthenticated: false } });
    wrapper = shallow(<ConnectedVCPrivateRoute store={store} />);
  });

  it("should show the isAuthenticated value when imported from store", () => {
    expect(wrapper.props().children.props.auth.isAuthenticated).toBe(false);
  });
});
