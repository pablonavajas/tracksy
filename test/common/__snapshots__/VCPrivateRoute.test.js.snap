// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VCPrivateRoute redirects to /startupPage, if isAuthenticated is true, isStartup is true and isLoading is false 1`] = `
<HashRouter>
  <Router
    history={
      Object {
        "action": "REPLACE",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/startupPage",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <VCPrivateRoute
      auth={
        Object {
          "isAuthenticated": true,
          "isLoading": false,
          "isStartup": true,
        }
      }
    >
      <Route
        render={[Function]}
      >
        <Redirect
          to="/startupPage"
        >
          <Lifecycle
            onMount={[Function]}
            onUpdate={[Function]}
            to="/startupPage"
          />
        </Redirect>
      </Route>
    </VCPrivateRoute>
  </Router>
</HashRouter>
`;

exports[`VCPrivateRoute redirects to login, if isAuthenticated is false  1`] = `
<HashRouter>
  <Router
    history={
      Object {
        "action": "REPLACE",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/login",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <VCPrivateRoute
      auth={
        Object {
          "isAuthenticated": false,
          "isLoading": false,
          "isStartup": false,
        }
      }
    >
      <Route
        render={[Function]}
      >
        <Redirect
          to="/login"
        >
          <Lifecycle
            onMount={[Function]}
            onUpdate={[Function]}
            to="/login"
          />
        </Redirect>
      </Route>
    </VCPrivateRoute>
  </Router>
</HashRouter>
`;

exports[`VCPrivateRoute shows preloader, if isLoading is true  1`] = `
<HashRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <VCPrivateRoute
      auth={
        Object {
          "isAuthenticated": true,
          "isLoading": true,
          "isStartup": false,
        }
      }
    >
      <Route
        render={[Function]}
      >
        <Preloader>
          <div>
            <div
              className="progress"
            >
              <div
                className="indeterminate"
              />
            </div>
          </div>
        </Preloader>
      </Route>
    </VCPrivateRoute>
  </Router>
</HashRouter>
`;
